### PR TITLE
Implement Token Gating

### PR DESCRIPTION
**Pull request by chrispader**
relates to: https://github.com/status-im/status-mobile/issues/14461

### Summary

Implements the Token Gating component, as per design here:
[https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Mobile-Components?node-id=2463%3A39587](https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Mobile-Components?node-id=2463%3A39587)

#### Platforms

- Android
- iOS

##### Functional

- public chats
- group chats
- networks

### Steps to test

- Open Status
- Login
- Try to join community/channel

### CODE REVIEW

**Comment by chrispader** [comment]
Current status with mock data in quo2 sandbox:

Simulator Screen Shot - iPhone 13 - 2022-09-05 at 14 29 44

**Comment by chrispader** [comment]
Waiting for the channel avatar component to get merged...

**Comment by chrispader** [comment]
Updated channel avatar and made button dynamic. This is how the token gating drawer looks like currently, with

insufficient tokens:

Simulator Screen Shot - iPhone 14 Pro - 2022-09-29 at 19 17 19

and with sufficient tokens:

Simulator Screen Shot - iPhone 14 Pro - 2022-09-29 at 19 17 49

File: .gitignore [hunk 1]
```diff
@@ -175,3 +175,5 @@ test/appium/tests/users.py
 ##node bindings
 /bin/
 /lib/
+
+/.calva/
```
Comment (J-Son89) [inline .gitignore:183]: should this be here? 🤨

File: ios/Podfile.lock [hunk 2]
```diff
@@ -238,7 +238,7 @@ PODS:
     - React
   - react-native-status (1.0.0):
     - React
-  - react-native-status-keycard (2.5.37):
+  - react-native-status-keycard (2.5.38):
```
Comment (J-Son89) [inline ios/Podfile.lock]: also needed?

File: src/quo2/components/buttons/button.cljs [hunk 3]
```diff
@@ -66,7 +66,12 @@
                               :label                {:style    {:color colors/white}}
                               :background-color     {:default  colors/neutral-95
                                                      :pressed  colors/neutral-95
-                                                     :disabled colors/neutral-95}}}
+                                                     :disabled colors/neutral-95}}
+           :purple    {:icon-color       colors/white
+                       :label            {:style {:color colors/white}}
+                       :background-color {:default  (get-in colors/customization [:purple 50])
+                                          :pressed  (get-in colors/customization [:purple 50])
```
Comment (J-Son89) [inline src/quo2/components/buttons/button.cljs]: I think you should use `custom-color` function here instead.
It's in the colors file.
e.g
```
(colors/custom-color :purple 50)
```

File: src/quo2/components/buttons/button.cljs [hunk 4]
```diff
@@ -124,7 +129,12 @@
                                                     :disabled colors/white-opa-5}}
            :shell            {:icon-color          colors/white
                               :label               {:style   {:color colors/white}}
-                              :background-color    {:default colors/neutral-95}}}})
+                              :background-color    {:default colors/neutral-95}}
+           :purple    {:icon-color       colors/white
+                       :label            {:style {:color colors/white}}
+                       :background-color {:default  (get-in colors/customization [:purple 50])
```
Comment (J-Son89) [inline src/quo2/components/buttons/button.cljs]: same here

File: src/quo2/components/community/token_gating.cljs [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+                                                          true
+                                                          token-requirement-list))
+
+(defn token-requirement-list [props]
+  (let [{:keys [gate]} props
+        [gate-type token-requirement-list] gate
+        is-sufficient? (is-token-requirement-met? token-requirement-list)
+        message (if
+                 (= gate-type :join)
+                  (str "Join requirements" (when-not is-sufficient? " not") " met")
```
Comment (J-Son89) [inline src/quo2/components/community/token_gating.cljs]: I think these strings should be wrapped in i18n labels and the string to the en.json files

File: src/quo2/components/community/token_gating.cljs [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+                  (if (= gate-type :read)
+                    (str "View only requirements" (when-not is-sufficient? " not") " met")
+                    (str "View and post requirements" (when-not is-sufficient? " not") " met")))]
```
Comment (J-Son89) [inline src/quo2/components/community/token_gating.cljs]: same here

File: src/quo2/components/community/token_gating.cljs [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+                      [rn/view {:style (when-not (= token-index 0) (get styles :token-tag-spacing))}
+                       [token-tag/token-tag {:token (:token token)
+                                             :value (:amount token)
+                                             :size :big
+                                             :is-required (:is-sufficient token)
+                                             :is-purchasable (:is-purchasable token)
+                                             :token-img-src
+                                             (js/require
+                                              (str
+                                               "../resources/images/tokens/mainnet/"
```
Comment (J-Son89) [inline src/quo2/components/community/token_gating.cljs]: hmm this seems like it should not be done this way. 

Perhaps you can move this js/require into the screens file so we can use it for the examples and just pass the `token-img-src` as a prop?

File: src/quo2/components/community/token_gating.cljs [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+(defn token-gating [props]
+  (let [{:keys [community channel]} props
+        type (if (some? community) :community :channel)
+        inner-props (if (= type :community) community channel)
+        name (:name inner-props)
+        emoji (:emoji inner-props)
+        emoji-background-color (:emoji-background-color inner-props)
+        community-avatar (:community-avatar inner-props)
+        gates (:gates inner-props)
+        lock-status (if
```
Comment (J-Son89) [inline src/quo2/components/community/token_gating.cljs:98]: I updated channel avatar component recently and it now uses `locked?` prop instead of `lock-status`. Can you rebase and update accordingly? :D

File: .gitignore (hunk 1 above)
Comment (chrispader) [reply .gitignore:183]: I don't know if it should be in this PR or added somewhere else, but VS Code and Calva are producing some local output files when connected to the REPL.

Should we add this to .gitignore in another PR, @cammellos / @jakubgs  ?

File: src/quo2/components/community/token_gating.cljs [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+                     (when-not (= gate-index 0)
+                       [rn/view {:style (get styles :divider)}])
+                     [token-requirement-list {:gate gate}]]) gates)
+
+     (when (= type :channel)
+       [:<>
+        [button/button {:type :purple :style (get styles :enter-button) :disabled lock-status} "# Enter channel"]
+        [text/text
+         {:size :paragraph-2 :style (get styles :info-text)}
+         "Entering channel will reveal your public addresses to the node owner"]])]))
```
Comment (J-Son89) [inline src/quo2/components/community/token_gating.cljs]: I think these strings should also be i18n labels

File: ios/Podfile.lock (hunk 2 above)
Comment (chrispader) [reply ios/Podfile.lock]: yes, this adds support for Xcode 14 and iOS 16. https://github.com/status-im/react-native-status-keycard/pull/46

File: src/quo2/components/buttons/button.cljs (hunk 3 above)
Comment (chrispader) [reply src/quo2/components/buttons/button.cljs]: Thanks! :) Didn't see that function

File: src/quo2/components/community/token_gating.cljs (hunk 5 above)
Comment (chrispader) [reply src/quo2/components/community/token_gating.cljs]: Thanks!

File: src/quo2/components/community/token_gating.cljs (hunk 9 above)
Comment (chrispader) [reply src/quo2/components/community/token_gating.cljs]: done!

File: src/quo2/components/community/token_gating.cljs (hunk 7 above)
Comment (chrispader) [reply src/quo2/components/community/token_gating.cljs]: Do you think? Otherwise we would have to pass the `token` and `token-img-src` props to the gate's token requirements.

Shouldn't this be as generic as possible?

Comment (J-Son89) [reply src/quo2/components/community/token_gating.cljs]: I might be missing context on this, but is it just for tokens that we have in the `resources` folder or is possible that tokens (more so their images) are loaded in from some endpoint?

File: .gitignore (hunk 1 above)
Comment (jakubgs) [reply .gitignore:183]: Why another PR? Just do it in a separate commit and don't squash it.

File: src/quo2/components/buttons/button.cljs [hunk 10]
```diff
@@ -124,7 +129,12 @@
                                                     :disabled colors/white-opa-5}}
            :shell            {:icon-color          colors/white
                               :label               {:style   {:color colors/white}}
-                              :background-color    {:default colors/neutral-95}}}})
+                              :background-color    {:default colors/neutral-95}}
+           :purple    {:icon-color       colors/white
+                       :label            {:style {:color colors/white}}
+                       :background-color {:default  (colors/custom-color :purple 50)
+                                          :pressed  (colors/custom-color :purple 50)
+                                          :disabled (colors/custom-color :purple 50)}}}})
```
Comment (Parveshdhull) [inline src/quo2/components/buttons/button.cljs]: Hi @chrispader, thank you for your work
1. For dark it should be purple 60
2. This only covers a single case, when community theme color is purple. We need to find a better way for doing this.

File: src/quo2/components/community/token_gating.cljs [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+            [quo2.foundations.colors :as colors]
+            [quo.theme :as theme]
+            [status-im.i18n.i18n :as i18n]))
+
+(def styles {:container {:flex 1}
+             :header-container {:flex-direction :row
+
+                                :align-items :center}
+             :header-spacing-community {:margin-bottom 16}
+             :header-spacing-channel {:margin-bottom 24}
```
Comment (Parveshdhull) [inline src/quo2/components/community/token_gating.cljs]: Please align map key-values ([ref](https://github.com/bbatsov/clojure-style-guide#map-keys-alignment))

File: src/quo2/components/community/token_gating.cljs [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+                       :margin-horizontal -20
+                       :margin-top 12
+                       :margin-bottom 8}
+             :enter-button {:margin-top 16}
+             :info-text {:margin-top 12
+                         :text-align :center
+                         :padding-horizontal 20}})
+
+(defn is-token-requirement-met? [token-requirement-list] (reduce
+                                                          #(or (and %1 (:is-sufficient %2)) false)
```
Comment (Parveshdhull) [inline src/quo2/components/community/token_gating.cljs]: 1. Please move reduce to next line
2. can we use `(fn [var1 var2]` instead so it would be more clear what `%1` and `%2` are.
3. #(or ... false) ? (isn't or and false are redundant?

File: src/quo2/components/community/token_gating.cljs [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+                                               ".png"))}]]])
+                   token-requirement-list)]]))
+
+(defn is-community-locked? [community]
+  (let [{gates :gates} community]
+    (or
+     (when (some? gates)
+       (when (vector? (:join gates))
+         (not (is-token-requirement-met? (:join gates)))))
+     false)))
```
Comment (Parveshdhull) [inline src/quo2/components/community/token_gating.cljs]: Same here, `(or ... false)`?

File: src/quo2/components/community/token_gating.cljs [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+                      :main-icons2/unlocked)
+          {:style (get styles :header-title-lock)
+           :color           (if dark?
+                              colors/neutral-40
+                              colors/neutral-50)}])]
+
+      [button/button {:type :outline
+                      :size 32
+                      :icon :info
+                      :style (get styles :header-info-button)} "info"]]
```
Comment (Parveshdhull) [inline src/quo2/components/community/token_gating.cljs]: Please don't hard code string, use `(i18n/label :t/info)` instead

File: src/quo/react_native.cljs [hunk 15]
```diff
@@ -173,3 +173,5 @@
      :y         y
      :height    height
      :width     width}))
+
+(defn alert [props] (.alert (.-Alert  rn) props))
```
Comment (flexsurfer) [inline src/quo/react_native.cljs]: we have show-popup in utils

File: src/quo2/components/buttons/button.cljs (hunk 10 above)
Comment (chrispader) [reply src/quo2/components/buttons/button.cljs]: Oh ok, didn't know, that this would be "community themed". Will change the schema for token gating, so a `community-color` can be defined :)

Comment (chrispader) [reply src/quo2/components/buttons/button.cljs]: But the button component color would also need to be adjustable...

Comment (chrispader) [reply src/quo2/components/buttons/button.cljs]: Will only specific (pre-defined) "theme colors" for communities be allowed, or can any color be chosen?

File: src/quo2/components/community/token_gating.cljs (hunk 11 above)
Comment (chrispader) [reply src/quo2/components/community/token_gating.cljs]: Should all map values be at the same column as well?

Comment (chrispader) [reply src/quo2/components/community/token_gating.cljs]: It appears, `make lint-fix` and Calva Formatting doesn't fix this automatically...

File: src/quo2/components/buttons/button.cljs (hunk 10 above)
Comment (Parveshdhull) [reply src/quo2/components/buttons/button.cljs]: any color ([ref](https://discord.com/channels/624205794384281629/626744384565674034/1019905254134911016))

File: src/quo2/components/community/token_gating.cljs (hunk 12 above)
Comment (chrispader) [reply src/quo2/components/community/token_gating.cljs]: 1. ✅
2. sure!
3. If i understand correctly, `and` returns the value of what is passed to it, also if it's `nil` for example...

>  If a form returns logical false (nil or false), and returns that value and doesn't evaluate any of the other expressions, otherwise it returns the value of the last expr. (and) returns true.

I added `(or ... false)` so that `is-token-requirement-met?` function definitely always returns a boolean.

What do you think?

File: src/quo2/components/buttons/button.cljs (hunk 10 above)
Comment (chrispader) [reply src/quo2/components/buttons/button.cljs]: thanks!

File: src/quo/react_native.cljs (hunk 15 above)
Comment (chrispader) [reply src/quo/react_native.cljs]: thanks! didn't know that :)

File: src/quo2/components/community/token_gating.cljs (hunk 12 above)
Comment (Parveshdhull) [reply src/quo2/components/community/token_gating.cljs]: I am not sure why you need boolean, nil is also a logical false. but anyway you can use something like this
```
(every?
 (fn [token]
   (get-in token [1 :is-sufficient]))
 token-requirement-list)
```

File: src/quo2/components/community/token_gating.cljs (hunk 11 above)
Comment (Parveshdhull) [reply src/quo2/components/community/token_gating.cljs]: > Should all map values be at the same column as well?

I am not sure what you mean. do you mean nested maps? Its up to you, but the main point is to make the code more readable, for example

image

File: src/quo2/components/community/token_gating.cljs (hunk 12 above)
Comment (chrispader) [reply src/quo2/components/community/token_gating.cljs]: Thanks a lot! I'm such a Clojure noob still, missing so many useful clojure functions 😅

File: src/quo2/components/community/token_gating.cljs [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+                      :main-icons2/unlocked)
+          {:style (get styles :header-title-lock)
+           :color           (if dark?
+                              colors/neutral-40
+                              colors/neutral-50)}])]
+
+      [button/button {:type :outline
+                      :size 32
+                      :icon true
+                      :style (get styles :header-info-button)} :info]]
```
Comment (Parveshdhull) [inline src/quo2/components/community/token_gating.cljs]: > `:info`

Hi, Sorry if I was not clear, but I didn't meant that we should use keyword instead of string. I meant we can put this string in `/translations/en.json` and then use it like `(i18n/label :t/info)`. So this string can be translated for other languages.

Comment (chrispader) [reply src/quo2/components/community/token_gating.cljs]: I know, but i noticed that i was using icon buttons wrong, since i don't need to provide a label (string) at all 😃

Comment (Parveshdhull) [reply src/quo2/components/community/token_gating.cljs]: oh I see. So second variable/label is not optional? (maybe we can make it optional?)

image

Comment (Parveshdhull) [reply src/quo2/components/community/token_gating.cljs]: ok I see now, second variable in this case is icon instead of label. So don't we need to use it like `:main-icons2/icon`?

Comment (chrispader) [reply src/quo2/components/community/token_gating.cljs]: if i understood it correctly, if `:icon true` is set, instead of a label the icon name is provided as the second argument

Comment (chrispader) [reply src/quo2/components/community/token_gating.cljs]: Label would need to be provided, if there should be text inside the button

Comment (chrispader) [reply src/quo2/components/community/token_gating.cljs]: yes probably, is adding `:main-icons2/...` to the icon name just a convention or is this technically necessary?

Comment (Parveshdhull) [reply src/quo2/components/community/token_gating.cljs]: I am not sure if it is necessary for quo2 button (maybe @flexsurfer will know better), but in general it is not just a convention. If you use `:main-icons/info` then file/image with `:info` name will be used as icon. But if you use `:main-icons2/info` then file/image with `:info+size` name will be used as icon. So in this case `:info32@2x.png`. ([ref](https://github.com/status-im/status-mobile/blob/8dd75205f359d0f8b3cf02e5cf8e662812fae94f/src/quo2/components/icon.cljs#L10))

Comment (flexsurfer) [reply src/quo2/components/community/token_gating.cljs]: namespace is not used, we cut it, its only for IDE, but quo2 button uses quo2 icons which adds sizes to the icon name

Comment (chrispader) [reply src/quo2/components/community/token_gating.cljs]: Oh ok, got it. So technically adding the namespace doesn't change anything. The only difference is in using button component from either quo or quo2. Thanks!

File: src/quo2/components/community/token_gating.cljs [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+                                              (str
+                                               "../resources/images/tokens/mainnet/"
+                                               (:token token)
+                                               ".png"))}]]])
+                   token-requirement-list)]]))
+
+(defn is-community-locked? [community]
+  (let [{gates :gates} community]
+    (when (some? gates)
+      (when (vector? (:join gates))
```
Comment (Parveshdhull) [inline src/quo2/components/community/token_gating.cljs]: you can just use `(when (vector? (:join gates))` only, `(when (some? gates)` is redundant

Comment (chrispader) [reply src/quo2/components/community/token_gating.cljs]: thanks!

File: src/quo2/components/community/token_gating.cljs [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+  (let [{gates :gates} community]
+    (when (vector? (:join gates)
+                   (not (is-token-requirement-met? (:join gates)))))))
+
+(defn is-channel-locked? [channel]
+  (let [{gates :gates} channel]
+    (if (vector? (:read gates))
+      (not (is-token-requirement-met? (:read gates)))
+      (when (vector? (:write gates))
+        (not (is-token-requirement-met? (:write gates)))))))
```
Comment (Parveshdhull) [inline src/quo2/components/community/token_gating.cljs]: we need to check every time if it is vector, while calling `is-token-requirement-met?` function, maybe better to move this check to function itself.

Comment (Parveshdhull) [reply src/quo2/components/community/token_gating.cljs]: so, this will become
```
(defn is-channel-locked? [channel]
  (let [{gates :gates} channel]
    (not (is-token-requirement-met? (or (:read gates) (:write gates)))
```

Comment (chrispader) [reply src/quo2/components/community/token_gating.cljs]: Checking for `vector?` is necessary, because i want to check which "gates" are set. Moving this to the enclosing function would require us to add state, since the gates from the props can change

Comment (chrispader) [reply src/quo2/components/community/token_gating.cljs]: if no `:read` gate is given, then this function will check if there is a `:write` gate...

Comment (Parveshdhull) [reply src/quo2/components/community/token_gating.cljs]: current implementation will return true for both `vector?` and `is-token-requirement-met?` if empty vector is passed `[]`

You can do something like
```
(defn is-token-requirement-met? [token-requirement-list]
  (and (seq token-requirement-list)
       (every?
        (fn [token]
          (get token :is-sufficient?))
        token-requirement-list)))
```
> if no :read gate is given, then this function will check if there is a :write gate...

And, for `is-channel-locked?`
```
(defn is-channel-locked? [channel]
  (let [{gates :gates} channel]
    (not (is-token-requirement-met? (or (:read gates) (:write gates)))   <------ This will make sure to pass whichever is available
```

Comment (Parveshdhull) [reply src/quo2/components/community/token_gating.cljs]: updated the function with `and`

Comment (chrispader) [reply src/quo2/components/community/token_gating.cljs]: thanks! 👍

File: src/quo2/components/buttons/button.cljs (hunk 10 above)
Comment (chrispader) [reply src/quo2/components/buttons/button.cljs]: Waiting for a approach from the design team on that, so this is still WIP

**Comment by chrispader** [comment]
This is what a very complex variant of this component could look like:

Simulator Screen Shot - iPhone 14 Pro - 2022-10-18 at 15 51 22

**Comment by chrispader** [comment]
@pedro-et @Parveshdhull we would need to add the `locked` icon with size 20 to the resources, otherwise using the icon doesn't work. For testing purposes i just exported it from Figma, but i think the format isn't 100% correct

File: src/quo/components/bottom_sheet/style.cljs [hunk 19]
```diff
@@ -23,12 +25,13 @@
...
 
-(defn content-container
-  [window-height]
-  {:background-color        (:ui-background @colors/theme)
-   :border-top-left-radius  border-radius
-   :border-top-right-radius border-radius
-   :height                  (* window-height 2)})
+(defn content-container  [window-height]
+  (let [dark? (theme/dark?)]
+    {:background-color (if dark? colors2/neutral-90 colors/white)
```
Comment (Parveshdhull) [inline src/quo/components/bottom_sheet/style.cljs]: hi @chrispader 
1. we should not change `quo` files
2. please use colors2 only. Here [themes-colors](https://github.com/status-im/status-mobile/blob/180df967e81c9c417e71db19670d1c64b56b2cfb/src/quo2/foundations/colors.cljs#L189) function can be used.
3. Also, if possible please do implementation and integration of component in separate PRs (easy to review)

File: src/quo2/components/community/token_gating.cljs [hunk 20]
```diff
@@ -0,0 +1,215 @@
...
+                    (i18n/label
+                     :t/community-channel-read-requirements {:not not-text})
+                    (i18n/label
+                     :t/community-channel-write-requirements {:not not-text})))]
+    [rn/view
+     [rn/view {:style (get styles :token-requirement-text-spacing)}
+      [text/text {:weight :medium
+                  :number-of-lines 1} message]
+      [text/text {:size :paragraph-2
+                  :number-of-lines 1} "You must hodl:"]]
```
Comment (Parveshdhull) [inline src/quo2/components/community/token_gating.cljs]: 1. Please move this string to [translations](https://github.com/status-im/status-mobile/blob/develop/translations/en.json) file
2. Maybe designs are changed, but it is not "hodl" anymore. [design](https://www.figma.com/file/WQZcp6S0EnzxdTL4taoKDv/Design-System?node-id=2463%3A39587)

File: src/quo2/components/info/information_box.cljs [hunk 21]
```diff
@@ -56,7 +56,9 @@
   [{:keys [type closable? closed? id icon style button-label on-button-press on-close]} message]
   (let [background-color (get-color-by-type type :bg)
         border-color     (get-color-by-type type :border)
-        icon-color       (get-color-by-type type :icon)
+        icon-color       (case (name icon)
+                           "untrustworthy" :no-color
+                           :none (get-color-by-type type :icon))
```
Comment (Parveshdhull) [inline src/quo2/components/info/information_box.cljs]: Please can you elaborate why these changes are needed.

File: src/quo2/components/list_items/channel.cljs [hunk 22]
```diff
@@ -1,26 +1,72 @@
...
+                         :on-press (when locked?
+                                     #(>evt
+                                       [:bottom-sheet/show-sheet
+                                        {:content
+                                         (constantly [token-gating/token-gating
+                                                      {:channel {:name name
+                                                                 :community-color (colors/custom-color :pink 50)
+                                                                 :emoji emoji
+                                                                 :emoji-background-color channel-color
+                                                                 :on-enter-channel (fn [] (utils/show-popup "Entered channel" "Wuhuu!! You successfully entered the channel :)"))
```
Comment (Parveshdhull) [inline src/quo2/components/list_items/channel.cljs]: 1. > Entered channel" "Wuhuu!! You successfully entered the channel :)

testing?
2. Maybe create function for this dispatch and assign it to on-press (more readable)

File: src/quo2/components/info/information_box.cljs (hunk 21 above)
Comment (chrispader) [reply src/quo2/components/info/information_box.cljs]: Because otherwise the white content of the `untrustworthy` icon (exclamation mark) would get painted over.

I decided to specifically do this check in `information-box` component, because there shouldn't be the need to manually set something like `:no-color true` on the top level component and pass it down to the icon all the way.

What do you think?

File: src/quo/components/bottom_sheet/style.cljs (hunk 19 above)
Comment (chrispader) [reply src/quo/components/bottom_sheet/style.cljs]: The problem is, that the old `bottom-sheet` component has the wrong background-color, so the content doesn't match. Where should i change this instead?

Or should we create a new `quo2/bottom-sheet` component?

File: src/quo2/components/info/information_box.cljs (hunk 21 above)
Comment (Parveshdhull) [reply src/quo2/components/info/information_box.cljs]: Thank you for explanation. There might be other use cases so let's not hard-code for `untrustworthy`
Maybe we can do something like this
```
icon-color (if no-icon-color? :transparent (get-color-by-type type :icon))
```
And we can remove changes from info_message.cljs

File: src/quo/components/bottom_sheet/style.cljs (hunk 19 above)
Comment (Parveshdhull) [reply src/quo/components/bottom_sheet/style.cljs]: Yes, I think we can copy this component in quo2 (or in ui2?) and make changes.
cc: @flexsurfer

File: src/quo2/components/info/information_box.cljs (hunk 21 above)
Comment (chrispader) [reply src/quo2/components/info/information_box.cljs]: Done! 👍

Comment (chrispader) [reply src/quo2/components/info/information_box.cljs]: `:transparent` doesnt' quite work as an icon color. It will fallback to `(colors/theme-colors colors/neutral-100 colors/white)`

[https://github.com/chrispader/status-mobile/blob/893fc2c6db5101f653c2733edaf86482c83f8a45/src/quo2/components/icon.cljs#L23](https://github.com/chrispader/status-mobile/blob/893fc2c6db5101f653c2733edaf86482c83f8a45/src/quo2/components/icon.cljs#L23)

Comment (chrispader) [reply src/quo2/components/info/information_box.cljs]: I'd instead just pass the `:no-icon-color?` prop to `info-message` and then the icon

Comment (chrispader) [reply src/quo2/components/info/information_box.cljs]: Screenshot 2022-10-20 at 15 41 10

File: src/quo/components/bottom_sheet/style.cljs (hunk 19 above)
Comment (flexsurfer) [reply src/quo/components/bottom_sheet/style.cljs]: yeah but not in this PR

File: ios/Podfile.lock [hunk 23]
```diff
@@ -640,7 +640,7 @@ SPEC CHECKSUMS:
   FBLazyVector: 3bb422f41b18121b71783a905c10e58606f7dc3e
   FBReactNativeSpec: f2c97f2529dd79c083355182cc158c9f98f4bd6e
   Folly: b73c3869541e86821df3c387eb0af5f65addfab4
-  glog: 36ce0530c6d2c3a5a4326885ef4069564887a1db
+  glog: 997518ea2aa2d8cd5df9797b641b758d52ecf2bc
```
Comment (ibrkhalil) [inline ios/Podfile.lock]: Do we need this change ?

File: src/quo2/components/community/token_gating.cljs [hunk 24]
```diff
@@ -0,0 +1,260 @@
...
+            [quo2.components.buttons.button :as button]
+            [quo2.components.icon :as icon]
+            [quo2.components.info.information-box :as information-box]
+            [quo2.components.markdown.text :as text]
+            [quo2.components.tags.token-tag :as token-tag]
+            [quo2.foundations.colors :as colors]
+            [status-im.i18n.i18n :as i18n]
+            [status-im.ui.components.fast-image :as fast-image]))
+
+(def token-tag-horizontal-spacing 7)
```
Comment (ibrkhalil) [inline src/quo2/components/community/token_gating.cljs]: Can be private
`(def token-tag-horizontal-spacing 7)`
to
`(def ^:private token-tag-horizontal-spacing 7)`

File: src/quo2/components/community/token_gating.cljs [hunk 25]
```diff
@@ -0,0 +1,260 @@
...
+   (map-indexed (fn [token-index token]
+                  (let [{:keys [token-img-src token amount is-sufficient? is-purchasable?]} token]
+                    ^{:key token-index}
+                    [rn/view {:style (get styles :token-tag-spacing)}
+                     [token-tag/token-tag {:token          token
+                                           :value          amount
+                                           :size           24
+                                           :border-color   community-color
+                                           :is-required    is-sufficient?
+                                           :is-purchasable is-purchasable?
```
Comment (ibrkhalil) [inline src/quo2/components/community/token_gating.cljs:80]: Can we add a question mark at the end to indicate it's a boolean ?
`:is-required` -> `:is-required?`

File: src/quo2/components/community/token_gating.cljs [hunk 26]
```diff
@@ -0,0 +1,260 @@
...
+                        :t/enter-channel))]
+            [text/text
+             {:size :paragraph-2
+              :style (merge
+                      (get styles :info-text)
+                      {:color (colors/theme-colors
+                               colors/neutral-50
+                               colors/neutral-40)})}
+             (i18n/label
+              :t/community-enter-channel-info)]]))])))
```
Comment (ibrkhalil) [inline src/quo2/components/community/token_gating.cljs]: Kindly add a new line @EOF 😄

File: src/quo2/screens/community/token_gating.cljs [hunk 27]
```diff
@@ -0,0 +1,165 @@
...
+                        :right 0
+                        :top 0}}
+       [preview/customizer state descriptor]]]
+     [rn/view {:height (if (= (:type @state) :community) 280 495) :margin-top 20}
+      [rn/view {:style (merge
+                        (get styles :container-sandbox)
+                        {:background-color (colors/theme-colors
+                                            colors/white
+                                            colors/neutral-90)})}
+       [quo2/token-gating preview-props]]]]))
```
Comment (ibrkhalil) [inline src/quo2/screens/community/token_gating.cljs]: New line at end of fille bitte ..

File: src/status_im/ui/screens/communities/community_overview_redesign.cljs [hunk 28]
```diff
@@ -1,21 +1,31 @@
...
+            [status-im.ui.screens.communities.request-to-join-bottom-sheet-redesign :as request-to-join]
+            [status-im.ui.screens.communities.styles :as styles]
+            [status-im.utils.handlers :refer [<sub >evt]]
+            [status-im.utils.utils :as utils]))
+
+(def knc-token-img (js/require "../resources/images/tokens/mainnet/KNC.png"))
+(def mana-token-img (js/require "../resources/images/tokens/mainnet/MANA.png"))
+(def rare-token-img (js/require "../resources/images/tokens/mainnet/RARE.png"))
+(def eth-token-img (js/require "../resources/images/tokens/mainnet/ETH.png"))
+(def dai-token-img (js/require "../resources/images/tokens/mainnet/DAI.png"))
```
Comment (ibrkhalil) [inline src/status_im/ui/screens/communities/community_overview_redesign.cljs]: Kindly fix the require list in the file, I think Calva automatically sorts requires and therefore making unnecessary changes to files

**Review by ibrkhalil — CHANGES_REQUESTED** [review]

File: ios/Podfile.lock (hunk 23 above)
Comment (chrispader) [reply ios/Podfile.lock]: This gets changed automatically for me once i run `make run-ios`...

File: src/quo2/components/community/token_gating.cljs (hunk 25 above)
Comment (chrispader) [reply src/quo2/components/community/token_gating.cljs:80]: That's implemented in `token_tag` component, not created by me. So i think this belongs into another PR

File: ios/Podfile.lock (hunk 23 above)
Comment (ibrkhalil) [reply ios/Podfile.lock]: Please revert

Comment (chrispader) [reply ios/Podfile.lock]: 👍

File: src/quo2/components/community/token_gating.cljs (hunk 25 above)
Comment (ibrkhalil) [reply src/quo2/components/community/token_gating.cljs:80]: Kindly read this

https://matheus.ro/2017/12/11/clean-code-boy-scout-rule/

Comment (chrispader) [reply src/quo2/components/community/token_gating.cljs:80]: All good, still i think these changes should be made in a separate PR. Other extra changes i made were requested to be implemented separately too.

**Review by ibrkhalil — APPROVED** [review]

**Comment by cammellos** [comment]
@chrispader apologies, would you mind rebasing one last time and then we can merge?
Thank you!

**Comment by chrispader** [comment]
> @chrispader apologies, would you mind rebasing one last time and then we can merge?
> Thank you!

done! 👍
